### ISSUE TITLE
Orchestrate startup based on components readiness

### ISSUE THREAD

**Issue by danielpetisme** [confluentinc/cp-ansible#415]
**For Confluent Enterprise Support customers, we would strongly advise you to open a Support ticket which will be addressed within your Support contract SLA at ```https://support.confluent.io```**

**Describe the issue**
Orchestrate confluent platform's components startup to prevent a global failing

**To Reproduce**
1- Install a Confluent Platform
2- Stop all services
3- Start al service in the same time
4- Brokers will crash because Zookeeper will not be live (ie. not ansering on 2181), the remaining components won't startup properly because brokers will be either starting but not live or stoppped.

**Expected behaviour**
The Confluent Platform's components should be able to retry the different connections to absorb the startup latency of each components (just as is done with the `kafka_broker_jolokia_urp_url` or other health checks).
A solution would be to provide an `ExecPreStart` script delaying the actual startup until  a set of conditions are fullfilled.
The Confluent Utility Belt is full of handy scripts to detect the readiness of the cluster components

**Inventory File**
N/A

**Logs**
N/A

**Environment (please complete the following information):**
N/A

**Additional context**

**Comment by JumaX** [comment]
Hello @danielpetisme Thank you for reaching out with your question.  It looks like you are speaking about Confluent Platform itself and not CP-Ansible.  Can you confirm this?

As you have mentioned CP-Ansible handles the orchestration of starting up various components, so I'm not sure it would make sense to include that in the platform itself.

If you have a support contract with Confluent, I would strongly advise you to open a support ticket with confluent support at https://support.confluent.io to discuss this further.

Please don't hesitate to let me know if I am misunderstanding your question.

Thanks

**Comment by danielpetisme** [comment]
I could imagine 2 options. Either is the software responsability to enforce readiness or it's cp-ansible concern by providing an appropriate `ExecPreStart` in the `override.conf` file.
What would be your recommendation ?

To give you context, we has an electrical power maintenance operation which shutdowned a non-prod environment. 
When the power was back on, the platform wasn't able to restart due to the lack of components startup dependencies.

I do have a contract support so I'll open a case.

**Comment by JumaX** [comment]
Hi Daniel,

Thanks for the context, this is very helpful.  In my opinion, it's not either CP-Ansible's responsibility or the Platforms responsibility to handle restarting the components after a power outage.  This is what the application administration team is responsible for orchestrating.  

The way I would approach this is as follows:

1. See if I could distribute the cluster over multiple DC's so as to never have it be completely down if their is such an outage

2. Write scripts in my Organizations orchestration tooling of choice (Ansible, Shell, Chef, etc) that run after an OS restart, and which start the components as outlined in our documentation linked above.

You could even take some of our Ansible work and modify it to accomplish this task, saving you some work.

Additional thoughts:

If you really need automatic recovery and failover with no intervention, I would investigate using kubernetes and Confluent Operator.  This again would be an even more powerful solution for fail over and fault tolerance.

https://docs.confluent.io/current/installation/operator/co-introduction.html

To summarize, I don't see this as being a fault in the Application itself or even in CP-Ansible, as it really comes down to the Administration team being able to orchestrate the environment they own based on the given environments characteristics.

**Comment by danielpetisme** [comment]
Hello Justin,

Personally I think it would be valuable to provide the best experience possible with the product and propose built-in resiliency of a cluster restart is a plus.
readiness and liveness are getting more and more hype in distributed systems to prevent failing to a lack of orchestration and more and more tools comes built-in implementation (basically wait and retries)
 
Now, I do not know which components should handle this responsibility however I see an opportunity for a quick-win and immediate implementation with cp-ansible and https://docs.confluent.io/current/installation/docker/development.html#confluent-platform-utility-belt-cub 

Regarding your proposals

1/ At the moment, this solution is not affordable given the low business criticity of the cluster
2/ This is what we are currently doing but this kind of local optimization prevent us to migrate from an home made Ansible recipe to cp-ansible. The objective is to use cp-ansible as-is, decorate it somehow maybe but avoid as much as possible to touch the internal code (thus maintain an internal version of cp-ansible).

**Comment by domenicbove** [comment]
@danielpetisme I really like the idea, but can you elaborate on the why you want this feature?

Your proposed solution with a `ExecPreStart` should totally work. But I don't know why you would reboot all of our components at once. If this is for os patching and rebooting, then you should go one host at a time to avoid a kafka outage.

Is this all about disaster recovery then?

Right now the operational logic of our components do not really fit into the constructs of systemd service files. That's why we built the k8s operator as Justin mention. And a lot of the operational logic of starting up components in the correct order is built into cp-ansible in fact. The way I see it there should be playbooks to handle your operational tasks, and thus we should go back to the use case.

**Comment by danielpetisme** [comment]
> Is this all about disaster recovery then?

yes, we had a power outage on a non-prod environment.

> Right now the operational logic of our components do not really fit into the constructs of systemd service files

I understand and agree with that (after digging deeper I understood the systemd dependencies where not made to check liveness but at the end, it works 😄 ). 

Let's rephrase the question, in a "non-k8s"/multi-dc environment (let's say a non-prod environment) how can Confluent help users to guarantee the cluster will restart properly without admin intervention.

**Comment by JumaX** [comment]
@danielpetisme Good question, which I think you may have already answered yourself. :). We do have our docker images, as you have pointed out, which do have checks in place to assist with an orderly restart.  I believe this would fit your use case perfectly.

1. The images are supplied and fully supported by Confluent

2. The images contain restart checks

3. There is no additional cost/overhead to running these docker images.  It would generally involve installing the docker runtime on your hosts, and then downloading and configuring the images for the environment.  We do have extensive documentation on configuration, as you have linked to previously.

The other alternative would be to have a professional services engagement to help write orchestration scripts for your environment.  These would be fully supported by our Professional services team.

I would strongly recommend opening a support ticket or reaching out to your account team to discuss these options further.

**Comment by danielpetisme** [comment]
I've opened a ticket on the confluent support portal.
I propose to close the issue, for transparency sake I'll try to update the current issue if anyone ask the same question in the future

**Comment by pantaoran** [comment]
I also have this problem. What was the outcome of your support ticket @danielpetisme ?
